File: app/src/main/java/org/digitalcampus/oppia/model/MultiLangInfoModel.java
/*
 * This file is part of OppiaMobile - https://digital-campus.org/
 *
 * OppiaMobile is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * OppiaMobile is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OppiaMobile. If not, see <http://www.gnu.org/licenses/>.
 */

package org.digitalcampus.oppia.model;

import android.content.SharedPreferences;
import android.text.TextUtils;
import android.util.Log;

import org.digitalcampus.oppia.activity.PrefsActivity;
import org.digitalcampus.oppia.analytics.Analytics;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;


public class MultiLangInfoModel implements Serializable {

    public static final String TAG = MultiLangInfoModel.class.getSimpleName();

    private List<Lang> langs = new ArrayList<>();
    private List<Lang> titles = new ArrayList<>();
    private List<Lang> descriptions = new ArrayList<>();

    public static final String DEFAULT_NOTITLE = "No title set";

    public String getTitle(String lang) {
        String title = getInfo(lang, titles);
        return title == null ? DEFAULT_NOTITLE : title.trim();
    }

    public String getTitle(SharedPreferences prefs){
        return this.getTitle(prefs.getString(PrefsActivity.PREF_LANGUAGE, Locale.getDefault().getLanguage()));
    }

    public void setTitles(List<Lang> titles) {
        this.titles = titles;
    }

    public void setTitlesFromJSONString(String jsonStr) {
        setInfoFromJSONString(jsonStr, this.titles, false);
    }

    public String getTitleJSONString(){
        return getInfoJSONString(this.titles);
    }

    public String getDescription(String lang) {
        return getInfo(lang, descriptions);
    }

    public String getDescription(SharedPreferences prefs){
        return this.getDescription(prefs.getString(PrefsActivity.PREF_LANGUAGE, Locale.getDefault().getLanguage()));
    }

    public void setDescriptions(List<Lang> descriptions) {
        this.descriptions = descriptions;
    }

    public void setDescriptionsFromJSONString(String jsonStr) {
        setInfoFromJSONString(jsonStr, this.descriptions, false);
    }

    public String getDescriptionJSONString(){
        return getInfoJSONString(this.descriptions);
    }

    public List<Lang> getLangs() {
        return langs;
    }
    public void setLangs(List<Lang> langs) {
        this.langs = langs;
    }

    public String getLangsJSONString(){
        return getInfoJSONString(this.langs);
    }

    public void setLangsFromJSONString(String jsonStr) {
        setInfoFromJSONString(jsonStr, this.langs, true);
    }

    private String getInfo(String lang, List<Lang> values){
        for(Lang l: values){
            if(l.getLanguage().equals(lang)){
                return l.getContent().trim();
            }
        }
        if(values.size() > 0){
            return values.get(0).getContent().trim();
        }

        return null;
    }

    private String getInfoJSONString(List<Lang> values){
        JSONArray array = new JSONArray();
        for(Lang l: values){
            JSONObject obj = new JSONObject();
            try {
                obj.put(l.getLanguage(), l.getContent());
            } catch (JSONException e) {
                Analytics.logException(e);
                Log.d(TAG, "JSON error: ", e);
            }
            array.put(obj);
        }
        return array.toString();
    }

    private void setInfoFromJSONString(String jsonStr, List<Lang> values, boolean isLangs){
        try {
            JSONArray infoArray = new JSONArray(jsonStr);
            for(int i=0; i< infoArray.length(); i++){
                JSONObject infoObj = infoArray.getJSONObject(i);
                @SuppressWarnings("unchecked")
                Iterator<String> iter = infoObj.keys();
                while(iter.hasNext()){
                    String key = iter.next();
                    String info = "";
                    if(!isLangs) {
                        info = infoObj.getString(key);
                    }
                    Lang l = new Lang(key, info);
                    values.add(l);
                }
            }
        } catch (JSONException e) {
            Analytics.logException(e);
            Log.d(TAG, "JSON error: ", e);
        } catch (NullPointerException npe){
            Analytics.logException(npe);
            Log.d(TAG, "Null pointer error: ", npe);
        }
    }

    public void setTitlesFromJSONObjectMap(JSONObject jsonObjectMultilang) throws JSONException {
        List<Lang> localLangs = parseLangs(jsonObjectMultilang);
        this.titles = localLangs;
    }

    public void setDescriptionsFromJSONObjectMap(JSONObject jsonObjectMultilang) throws JSONException {
        List<Lang> localLangs = parseLangs(jsonObjectMultilang);
        this.descriptions = localLangs;
    }

    private List<Lang> parseLangs(JSONObject jsonObjectMultilang) throws JSONException {
        Iterator<String> keys = jsonObjectMultilang.keys();
        List<Lang> localLangs = new ArrayList<>();

        while(keys.hasNext()) {
            String key = keys.next();
            String value = jsonObjectMultilang.getString(key);
            if (!TextUtils.isEmpty(value) && !TextUtils.equals(value,"null")){
                localLangs.add(new Lang(key, value));
            }
        }

        return localLangs;
    }
}

File: app/src/main/java/org/digitalcampus/oppia/model/MultiLangInfoModel.kt
/*
 * This file is part of OppiaMobile - https://digital-campus.org/
 *
 * OppiaMobile is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * OppiaMobile is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OppiaMobile. If not, see <http://www.gnu.org/licenses/>.
 */
package org.digitalcampus.oppia.model

import android.content.SharedPreferences
import android.util.Log
import org.digitalcampus.oppia.activity.PrefsActivity
import org.digitalcampus.oppia.analytics.Analytics
import org.digitalcampus.oppia.utils.TextUtilsJava
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.Serializable
import java.util.Locale

open class MultiLangInfoModel : Serializable {

    companion object {
        val TAG = MultiLangInfoModel::class.simpleName
        const val DEFAULT_NOTITLE = "No title set"
    }

    private var langs: MutableList<Lang> = ArrayList()
    private var titles: MutableList<Lang> = ArrayList()
    private var descriptions: MutableList<Lang> = ArrayList()

    fun getTitle(lang: String?): String {
        val title = getInfo(lang, titles)
        return title?.trim() ?: DEFAULT_NOTITLE
    }

    fun getTitle(prefs: SharedPreferences): String {
        return this.getTitle(
            prefs.getString(
                PrefsActivity.PREF_CONTENT_LANGUAGE,
                Locale.getDefault().language
            )
        )
    }

    fun setTitles(titles: MutableList<Lang>) {
        this.titles = titles
    }

    fun setTitlesFromJSONString(jsonStr: String) {
        setInfoFromJSONString(jsonStr, titles, false)
    }

    fun getTitleJSONString(): String {
        return getInfoJSONString(titles)
    }

    fun getDescription(lang: String?): String? {
        return getInfo(lang, descriptions)
    }

    fun getDescription(prefs: SharedPreferences): String? {
        return this.getDescription(
            prefs.getString(
                PrefsActivity.PREF_CONTENT_LANGUAGE,
                Locale.getDefault().language
            )
        )
    }

    fun setDescriptions(descriptions: MutableList<Lang>) {
        this.descriptions = descriptions
    }

    fun setDescriptionsFromJSONString(jsonStr: String) {
        setInfoFromJSONString(jsonStr, descriptions, false)
    }

    fun getDescriptionJSONString(): String {
        return getInfoJSONString(descriptions)
    }

    fun getLangs(): List<Lang> {
        return langs
    }

    fun setLangs(langs: MutableList<Lang>) {
        this.langs = langs
    }

    fun getLangsJSONString(): String {
        return getInfoJSONString(langs)
    }

    fun setLangsFromJSONString(jsonStr: String) {
        setInfoFromJSONString(jsonStr, langs, true)
    }

    private fun getInfo(lang: String?, values: List<Lang>): String? {
        for ((language, content) in values) {
            if (language.equals(lang, ignoreCase = true)) {
                return content.trim()
            }
        }
        return if (values.isNotEmpty()) {
            values[0].content.trim()
        } else null
    }

    private fun getInfoJSONString(values: List<Lang>): String {
        val array = JSONArray()
        for ((language, content) in values) {
            val obj = JSONObject()
            try {
                obj.put(language, content)
            } catch (e: JSONException) {
                Analytics.logException(e)
                Log.d(TAG, "JSON error: ", e)
            }
            array.put(obj)
        }
        return array.toString()
    }

    private fun setInfoFromJSONString(jsonStr: String, values: MutableList<Lang>, isLangs: Boolean) {
        try {
            val infoArray = JSONArray(jsonStr)
            for (i in 0 until infoArray.length()) {
                val infoObj = infoArray.getJSONObject(i)
                val iter = infoObj.keys()
                while (iter.hasNext()) {
                    val key = iter.next()
                    var info: String? = ""
                    if (!isLangs) {
                        info = infoObj.getString(key)
                    }
                    val l = Lang(key, info!!)
                    values.add(l)
                }
            }
        } catch (e: JSONException) {
            Analytics.logException(e)
            Log.d(TAG, "JSON error: ", e)
        } catch (npe: NullPointerException) {
            Analytics.logException(npe)
            Log.d(TAG, "Null pointer error: ", npe)
        }
    }

    @Throws(JSONException::class)
    fun setTitlesFromJSONObjectMap(jsonObjectMultilang: JSONObject) {
        val localLangs = parseLangs(jsonObjectMultilang)
        titles = localLangs
    }

    @Throws(JSONException::class)
    fun setDescriptionsFromJSONObjectMap(jsonObjectMultilang: JSONObject) {
        val localLangs = parseLangs(jsonObjectMultilang)
        descriptions = localLangs
    }

    @Throws(JSONException::class)
    private fun parseLangs(jsonObjectMultilang: JSONObject): MutableList<Lang> {
        val keys = jsonObjectMultilang.keys()
        val localLangs: MutableList<Lang> = ArrayList()
        while (keys.hasNext()) {
            val key = keys.next()
            val value = jsonObjectMultilang.getString(key)
            if (!TextUtilsJava.isEmpty(value) && !TextUtilsJava.equals(value, "null")) {
                localLangs.add(Lang(key, value))
            }
        }
        return localLangs
    }
}